File: packages/core/src/protocol/documents.ts
import { z } from "zod";
import { ProtocolError, type ProtocolErrorCode } from "./errors.js";
import {
  caseIdSchema,
  checkResultIdSchema,
  evalRunIdSchema,
  invocationIdSchema,
  nonBlankStringSchema,
  safeIntegerSchema,
  spanIdSchema,
  traceIdSchema,
  u32Schema,
} from "./identifiers.js";
import {
  invalidJsonPath,
  isJsonObject,
  isJsonValue,
  type JsonObject,
  type JsonValue,
} from "./json.js";
import { timestampSchema } from "./timestamp.js";

export const SCHEMA_VERSION = "kensa.protocol.v1" as const;
export const DOCUMENT_KINDS = [
  "eval_run",
  "invocation",
  "span",
  "check_result",
] as const;
const jsonValueSchema = z.json().refine(isJsonValue, {
  message: "Value is not JSON-compatible",
  params: { kensaCode: "invalid_json_value" },
}) as z.ZodType<JsonValue>;
const jsonObjectSchema = z.record(z.string(), z.json()).refine(isJsonObject, {
  message: "Value must be a JSON object",
  params: { kensaCode: "invalid_json_value" },
}) as z.ZodType<JsonObject>;
const failureCategorySchema = z.enum([
  "agent",
  "simulator",
  "judge",
  "configuration",
  "infrastructure",
  "harness",
  "unknown",
]);
const effectPolicySchema = z.enum(["none", "captured", "sandboxed", "live"]);
const evidenceStatusSchema = z.enum([
  "complete",
  "pending",
  "partial",
  "unavailable",
]);
const failureSchema = z.strictObject({
  category: failureCategorySchema,
  kind: nonBlankStringSchema,
  message: nonBlankStringSchema,
  evidence: jsonObjectSchema,
});
const provenanceSchema = z.strictObject({
  producer: nonBlankStringSchema,
  producer_version: nonBlankStringSchema,
  adapter: nonBlankStringSchema.nullable(),
  adapter_version: nonBlankStringSchema.nullable(),
  runtime: nonBlankStringSchema,
  runtime_version: nonBlankStringSchema,
  revision: nonBlankStringSchema.nullable(),
  environment: nonBlankStringSchema.nullable(),
  effects: effectPolicySchema,
});
const completenessSchema = z.strictObject({
  status: evidenceStatusSchema,
  reason: nonBlankStringSchema.nullable(),
});
const caseSnapshotSchema = z.strictObject({
  id: caseIdSchema,
  input: jsonValueSchema,
  metadata: jsonObjectSchema,
});
const common = { schema_version: z.literal(SCHEMA_VERSION) };

const evalRunBase = z.strictObject({
  ...common,
  document_kind: z.literal("eval_run"),
  id: evalRunIdSchema,
  status: z.enum([
    "pending",
    "running",
    "pass",
    "fail",
    "error",
    "cancelled",
    "interrupted",
  ]),
  created_at: timestampSchema,
  started_at: timestampSchema.nullable(),
  ended_at: timestampSchema.nullable(),
  duration_ms: safeIntegerSchema.nullable(),
  attributes: jsonObjectSchema,
  failure: failureSchema.nullable(),
});
const invocationBase = z.strictObject({
  ...common,
  document_kind: z.literal("invocation"),
  id: invocationIdSchema,
  run_id: evalRunIdSchema,
  case: caseSnapshotSchema,
  attempt: u32Schema.min(1),
  status: z.enum([
    "pending",
    "running",
    "pass",
    "fail",
    "error",
    "cancelled",
    "skipped",
    "interrupted",
  ]),
  started_at: timestampSchema.nullable(),
  ended_at: timestampSchema.nullable(),
  duration_ms: safeIntegerSchema.nullable(),
  output_recorded: z.boolean(),
  output: jsonValueSchema.nullable(),
  provenance: provenanceSchema,
  evidence_completeness: completenessSchema,
  attributes: jsonObjectSchema,
  failure: failureSchema.nullable(),
});
const spanBase = z.strictObject({
  ...common,
  document_kind: z.literal("span"),
  invocation_id: invocationIdSchema,
  trace_id: traceIdSchema,
  span_id: spanIdSchema,
  parent_span_id: spanIdSchema.nullable(),
  name: nonBlankStringSchema,
  span_kind: nonBlankStringSchema,
  status: z.enum(["unset", "ok", "error"]),
  status_message: nonBlankStringSchema.nullable(),
  started_at: timestampSchema.nullable(),
  ended_at: timestampSchema.nullable(),
  duration_ms: safeIntegerSchema.nullable(),
  input_recorded: z.boolean(),
  input: jsonValueSchema.nullable(),
  output_recorded: z.boolean(),
  output: jsonValueSchema.nullable(),
  attributes: jsonObjectSchema,
});
const checkResultBase = z.strictObject({
  ...common,
  document_kind: z.literal("check_result"),
  id: checkResultIdSchema,
  invocation_id: invocationIdSchema,
  name: nonBlankStringSchema,
  status: z.enum(["pass", "fail", "error", "skipped"]),
  started_at: timestampSchema.nullable(),
  ended_at: timestampSchema.nullable(),
  duration_ms: safeIntegerSchema.nullable(),
  evidence: jsonObjectSchema,
  failure: failureSchema.nullable(),
});

function withSemantics<T extends z.ZodType>(
  schema: T,
  check: (value: z.infer<T>, ctx: z.RefinementCtx) => void,
): T {
  return schema.superRefine(check);
}
const requiresFailure = (status: string) =>
  ["fail", "error", "cancelled", "interrupted", "skipped"].includes(status);
const evalRunSchema = withSemantics(evalRunBase, (value, ctx) => {
  if (requiresFailure(value.status) !== (value.failure !== null))
    ctx.addIssue({
      code: "custom",
      path: ["failure"],
      message: "Failure must match terminal status",
      params: { kensaCode: "contradictory_fields" },
    });
});
const invocationSchema = withSemantics(invocationBase, (value, ctx) => {
  if (requiresFailure(value.status) !== (value.failure !== null))
    ctx.addIssue({
      code: "custom",
      path: ["failure"],
      message: "Failure must match terminal status",
      params: { kensaCode: "contradictory_fields" },
    });
  if (!value.output_recorded && value.output !== null)
    ctx.addIssue({
      code: "custom",
      path: ["output"],
      message: "Unrecorded output must be null",
      params: { kensaCode: "contradictory_fields" },
    });
  if (
    value.evidence_completeness.status === "complete"
      ? value.evidence_completeness.reason !== null
      : value.evidence_completeness.reason === null
  )
    ctx.addIssue({
      code: "custom",
      path: ["evidence_completeness", "reason"],
      message: "Evidence reason contradicts status",
      params: { kensaCode: "contradictory_fields" },
    });
});
const spanSchema = withSemantics(spanBase, (value, ctx) => {
  if ((value.status === "error") !== (value.status_message !== null))
    ctx.addIssue({
      code: "custom",
      path: ["status_message"],
      message: "Error spans require a message",
      params: { kensaCode: "contradictory_fields" },
    });
  if (!value.input_recorded && value.input !== null)
    ctx.addIssue({
      code: "custom",
      path: ["input"],
      message: "Unrecorded input must be null",
      params: { kensaCode: "contradictory_fields" },
    });
  if (!value.output_recorded && value.output !== null)
    ctx.addIssue({
      code: "custom",
      path: ["output"],
      message: "Unrecorded output must be null",
      params: { kensaCode: "contradictory_fields" },
    });
});
const checkResultSchema = withSemantics(checkResultBase, (value, ctx) => {
  if ((value.status === "pass") !== (value.failure === null))
    ctx.addIssue({
      code: "custom",
      path: ["failure"],
      message: "Failure must match check status",
      params: { kensaCode: "contradictory_fields" },
    });
});

export {
  caseSnapshotSchema,
  completenessSchema,
  failureSchema,
  provenanceSchema,
  evalRunSchema,
  invocationSchema,
  spanSchema,
  checkResultSchema,
};
export const protocolDocumentSchema = z.discriminatedUnion("document_kind", [
  evalRunSchema,
  invocationSchema,
  spanSchema,
  checkResultSchema,
]);
type DeepReadonly<T> = T extends readonly (infer Item)[]
  ? readonly DeepReadonly<Item>[]
  : T extends object
    ? { readonly [Key in keyof T]: DeepReadonly<T[Key]> }
    : T;
export type EvalRun = DeepReadonly<z.infer<typeof evalRunSchema>>;
export type Invocation = DeepReadonly<z.infer<typeof invocationSchema>>;
export type Span = DeepReadonly<z.infer<typeof spanSchema>>;
export type CheckResult = DeepReadonly<z.infer<typeof checkResultSchema>>;
export type ProtocolDocument = DeepReadonly<
  z.infer<typeof protocolDocumentSchema>
>;
export type CaseSnapshot = DeepReadonly<z.infer<typeof caseSnapshotSchema>>;
export type Failure = DeepReadonly<z.infer<typeof failureSchema>>;
export type ExecutionProvenance = DeepReadonly<
  z.infer<typeof provenanceSchema>
>;
export type EvidenceCompleteness = DeepReadonly<
  z.infer<typeof completenessSchema>
>;
export type FailureCategory = z.infer<typeof failureCategorySchema>;
export type EffectPolicy = z.infer<typeof effectPolicySchema>;
export type EvidenceStatus = z.infer<typeof evidenceStatusSchema>;
export type EvalRunStatus = EvalRun["status"];
export type InvocationStatus = Invocation["status"];
export type SpanStatus = Span["status"];
export type CheckStatus = CheckResult["status"];

function errorCode(issue: z.core.$ZodIssue): ProtocolErrorCode {
  const params: unknown =
    "params" in issue
      ? (issue as { readonly params?: unknown }).params
      : undefined;
  const custom =
    typeof params === "object" && params !== null && "kensaCode" in params
      ? params.kensaCode
      : undefined;
  if (custom === "contradictory_fields" || custom === "invalid_json_value")
    return custom;
  const path = issue.path.map(String);
  if (path.some((part) => part.endsWith("_at") || part === "created_at"))
    return "invalid_timestamp";
  if (path.some((part) => part === "id" || part.endsWith("_id")))
    return "invalid_identifier";
  if (issue.code === "unrecognized_keys") return "unknown_field";
  if (
    issue.code === "invalid_value" ||
    path.at(-1) === "schema_version" ||
    path.at(-1) === "document_kind"
  )
    return "invalid_literal";
  if (
    path.at(-1) === "duration_ms" ||
    path.at(-1) === "attempt" ||
    issue.code === "too_big" ||
    issue.code === "too_small" ||
    issue.code === "not_multiple_of"
  )
    return "unsafe_integer";
  return "invalid_type";
}
function parse<T>(schema: z.ZodType<T>, value: unknown): T {
  const invalidPath = invalidJsonPath(value);
  if (invalidPath !== null) {
    const topLevelField = invalidPath.length === 1 ? invalidPath[0] : undefined;
    throw new ProtocolError(
      "runtime",
      topLevelField === "duration_ms" || topLevelField === "attempt"
        ? "unsafe_integer"
        : "invalid_json_value",
      invalidPath,
      "Value is not JSON-compatible",
    );
  }
  const result = schema.safeParse(value);
  if (result.success) return result.data;
  const issue = result.error.issues.reduce((first) => first);
  const issuePath =
    issue.code === "unrecognized_keys" &&
    "keys" in issue &&
    Array.isArray(issue.keys) &&
    issue.keys.length > 0
      ? [...issue.path, issue.keys[0]]
      : issue.path;
  throw new ProtocolError(
    "runtime",
    errorCode(issue),
    issuePath.map(String),
    issue.message,
  );
}
export const parseEvalRun = (value: unknown): EvalRun =>
  parse(evalRunSchema, value);
export const parseInvocation = (value: unknown): Invocation =>
  parse(invocationSchema, value);
export const parseSpan = (value: unknown): Span => parse(spanSchema, value);
export const parseCheckResult = (value: unknown): CheckResult =>
  parse(checkResultSchema, value);
export const parseProtocolDocument = (value: unknown): ProtocolDocument =>
  parse(protocolDocumentSchema, value);
export function parseProtocolJson(
  input: string | Uint8Array,
): ProtocolDocument {
  try {
    const text =
      typeof input === "string"
        ? input
        : new TextDecoder("utf-8", { fatal: true }).decode(input);
    return parseProtocolDocument(JSON.parse(text) as unknown);
  } catch (error) {
    if (error instanceof ProtocolError) throw error;
    throw new ProtocolError("syntax", "invalid_json", [], "Invalid JSON");
  }
}

File: crates/kensa-core/src/protocol/documents.rs
use std::collections::BTreeMap;
use std::num::NonZeroU32;

use schemars::JsonSchema;
use serde::{Deserialize, Deserializer, Serialize, Serializer, de, ser::SerializeStruct};
use serde_json::Value;

use super::primitives::{
    CaseId, CheckResultId, EvalRunId, InvocationId, NonBlankString, SpanId, Timestamp, TraceId,
};

pub type JsonObject = BTreeMap<String, Value>;

macro_rules! serialize_validated {
    ($name:ident { $($field:ident),+ $(,)? }) => {
        impl Serialize for $name {
            fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
            where
                S: Serializer,
            {
                self.validate().map_err(serde::ser::Error::custom)?;
                let field_count = [$(stringify!($field)),+].len();
                let mut state = serializer.serialize_struct(stringify!($name), field_count)?;
                $(state.serialize_field(stringify!($field), &self.$field)?;)+
                state.end()
            }
        }
    };
}

fn required_option<'de, D, T>(deserializer: D) -> Result<Option<T>, D::Error>
where
    D: Deserializer<'de>,
    T: Deserialize<'de>,
{
    Option::deserialize(deserializer)
}

macro_rules! wire_enum {
    ($name:ident { $($variant:ident),+ $(,)? }) => {
        #[derive(Clone, Copy, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
        #[serde(rename_all = "snake_case")]
        pub enum $name {
            $($variant),+
        }
    };
}

wire_enum!(EvalRunStatus {
    Pending,
    Running,
    Pass,
    Fail,
    Error,
    Cancelled,
    Interrupted,
});
wire_enum!(InvocationStatus {
    Pending,
    Running,
    Pass,
    Fail,
    Error,
    Cancelled,
    Skipped,
    Interrupted,
});
wire_enum!(CheckStatus {
    Pass,
    Fail,
    Error,
    Skipped,
});
wire_enum!(SpanStatus { Unset, Ok, Error });
wire_enum!(FailureCategory {
    Agent,
    Simulator,
    Judge,
    Configuration,
    Infrastructure,
    Harness,
    Unknown,
});
wire_enum!(EffectPolicy {
    None,
    Captured,
    Sandboxed,
    Live,
});
wire_enum!(EvidenceStatus {
    Complete,
    Pending,
    Partial,
    Unavailable,
});

#[derive(Clone, Copy, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
pub enum SchemaVersion {
    #[serde(rename = "kensa.protocol.v1")]
    V1,
}

macro_rules! document_kind {
    ($name:ident, $variant:ident, $wire:literal) => {
        #[derive(Clone, Copy, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
        pub enum $name {
            #[serde(rename = $wire)]
            $variant,
        }
    };
}

document_kind!(EvalRunDocumentKind, EvalRun, "eval_run");
document_kind!(InvocationDocumentKind, Invocation, "invocation");
document_kind!(SpanDocumentKind, Span, "span");
document_kind!(CheckResultDocumentKind, CheckResult, "check_result");

#[derive(Clone, Debug, Deserialize, JsonSchema, PartialEq, Serialize)]
#[serde(deny_unknown_fields)]
pub struct CaseSnapshot {
    pub id: CaseId,
    pub input: Value,
    pub metadata: JsonObject,
}

#[derive(Clone, Debug, Deserialize, JsonSchema, PartialEq, Serialize)]
#[serde(deny_unknown_fields)]
pub struct Failure {
    pub category: FailureCategory,
    pub kind: NonBlankString,
    pub message: NonBlankString,
    pub evidence: JsonObject,
}

#[derive(Clone, Debug, Deserialize, JsonSchema, PartialEq, Serialize)]
#[serde(deny_unknown_fields)]
pub struct ExecutionProvenance {
    pub producer: NonBlankString,
    pub producer_version: NonBlankString,
    #[serde(deserialize_with = "required_option")]
    pub adapter: Option<NonBlankString>,
    #[serde(deserialize_with = "required_option")]
    pub adapter_version: Option<NonBlankString>,
    pub runtime: NonBlankString,
    pub runtime_version: NonBlankString,
    #[serde(deserialize_with = "required_option")]
    pub revision: Option<NonBlankString>,
    #[serde(deserialize_with = "required_option")]
    pub environment: Option<NonBlankString>,
    pub effects: EffectPolicy,
}

#[derive(Clone, Debug, JsonSchema, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct EvidenceCompleteness {
    pub status: EvidenceStatus,
    pub reason: Option<NonBlankString>,
}

impl EvidenceCompleteness {
    pub fn validate(&self) -> Result<(), &'static str> {
        let valid = match self.status {
            EvidenceStatus::Complete => self.reason.is_none(),
            EvidenceStatus::Pending | EvidenceStatus::Partial | EvidenceStatus::Unavailable => {
                self.reason.is_some()
            }
        };
        if valid {
            Ok(())
        } else {
            Err("reason contradicts evidence completeness status")
        }
    }
}

serialize_validated!(EvidenceCompleteness { status, reason });

#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct RawEvidenceCompleteness {
    status: EvidenceStatus,
    #[serde(deserialize_with = "required_option")]
    reason: Option<NonBlankString>,
}

impl<'de> Deserialize<'de> for EvidenceCompleteness {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let raw = RawEvidenceCompleteness::deserialize(deserializer)?;
        let value = Self {
            status: raw.status,
            reason: raw.reason,
        };
        value.validate().map_err(de::Error::custom)?;
        Ok(value)
    }
}

#[derive(Clone, Debug, JsonSchema, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct EvalRun {
    pub schema_version: SchemaVersion,
    pub document_kind: EvalRunDocumentKind,
    pub id: EvalRunId,
    pub status: EvalRunStatus,
    pub created_at: Timestamp,
    pub started_at: Option<Timestamp>,
    pub ended_at: Option<Timestamp>,
    pub duration_ms: Option<u64>,
    pub attributes: JsonObject,
    pub failure: Option<Failure>,
}

impl EvalRun {
    pub fn validate(&self) -> Result<(), &'static str> {
        let failure_required = matches!(
            self.status,
            EvalRunStatus::Fail
                | EvalRunStatus::Error
                | EvalRunStatus::Cancelled
                | EvalRunStatus::Interrupted
        );
        if failure_required == self.failure.is_some() {
            Ok(())
        } else {
            Err("failure contradicts eval run status")
        }
    }
}

serialize_validated!(EvalRun {
    schema_version,
    document_kind,
    id,
    status,
    created_at,
    started_at,
    ended_at,
    duration_ms,
    attributes,
    failure,
});

#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct RawEvalRun {
    schema_version: SchemaVersion,
    document_kind: EvalRunDocumentKind,
    id: EvalRunId,
    status: EvalRunStatus,
    created_at: Timestamp,
    #[serde(deserialize_with = "required_option")]
    started_at: Option<Timestamp>,
    #[serde(deserialize_with = "required_option")]
    ended_at: Option<Timestamp>,
    #[serde(deserialize_with = "required_option")]
    duration_ms: Option<u64>,
    attributes: JsonObject,
    #[serde(deserialize_with = "required_option")]
    failure: Option<Failure>,
}

impl<'de> Deserialize<'de> for EvalRun {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let raw = RawEvalRun::deserialize(deserializer)?;
        let value = Self {
            schema_version: raw.schema_version,
            document_kind: raw.document_kind,
            id: raw.id,
            status: raw.status,
            created_at: raw.created_at,
            started_at: raw.started_at,
            ended_at: raw.ended_at,
            duration_ms: raw.duration_ms,
            attributes: raw.attributes,
            failure: raw.failure,
        };
        value.validate().map_err(de::Error::custom)?;
        Ok(value)
    }
}

#[derive(Clone, Debug, JsonSchema, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct Invocation {
    pub schema_version: SchemaVersion,
    pub document_kind: InvocationDocumentKind,
    pub id: InvocationId,
    pub run_id: EvalRunId,
    pub case: CaseSnapshot,
    pub attempt: NonZeroU32,
    pub status: InvocationStatus,
    pub started_at: Option<Timestamp>,
    pub ended_at: Option<Timestamp>,
    pub duration_ms: Option<u64>,
    pub output_recorded: bool,
    pub output: Option<Value>,
    pub provenance: ExecutionProvenance,
    pub evidence_completeness: EvidenceCompleteness,
    pub attributes: JsonObject,
    pub failure: Option<Failure>,
}

impl Invocation {
    pub fn validate(&self) -> Result<(), &'static str> {
        let failure_required = matches!(
            self.status,
            InvocationStatus::Fail
                | InvocationStatus::Error
                | InvocationStatus::Cancelled
                | InvocationStatus::Skipped
                | InvocationStatus::Interrupted
        );
        if failure_required != self.failure.is_some() {
            return Err("failure contradicts invocation status");
        }
        if !self.output_recorded && self.output.is_some() {
            return Err("output present when output_recorded is false");
        }
        self.evidence_completeness.validate()
    }
}

serialize_validated!(Invocation {
    schema_version,
    document_kind,
    id,
    run_id,
    case,
    attempt,
    status,
    started_at,
    ended_at,
    duration_ms,
    output_recorded,
    output,
    provenance,
    evidence_completeness,
    attributes,
    failure,
});

#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct RawInvocation {
    schema_version: SchemaVersion,
    document_kind: InvocationDocumentKind,
    id: InvocationId,
    run_id: EvalRunId,
    case: CaseSnapshot,
    attempt: NonZeroU32,
    status: InvocationStatus,
    #[serde(deserialize_with = "required_option")]
    started_at: Option<Timestamp>,
    #[serde(deserialize_with = "required_option")]
    ended_at: Option<Timestamp>,
    #[serde(deserialize_with = "required_option")]
    duration_ms: Option<u64>,
    output_recorded: bool,
    #[serde(deserialize_with = "required_option")]
    output: Option<Value>,
    provenance: ExecutionProvenance,
    evidence_completeness: EvidenceCompleteness,
    attributes: JsonObject,
    #[serde(deserialize_with = "required_option")]
    failure: Option<Failure>,
}

impl<'de> Deserialize<'de> for Invocation {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let raw = RawInvocation::deserialize(deserializer)?;
        let value = Self {
            schema_version: raw.schema_version,
            document_kind: raw.document_kind,
            id: raw.id,
            run_id: raw.run_id,
            case: raw.case,
            attempt: raw.attempt,
            status: raw.status,
            started_at: raw.started_at,
            ended_at: raw.ended_at,
            duration_ms: raw.duration_ms,
            output_recorded: raw.output_recorded,
            output: raw.output,
            provenance: raw.provenance,
            evidence_completeness: raw.evidence_completeness,
            attributes: raw.attributes,
            failure: raw.failure,
        };
        value.validate().map_err(de::Error::custom)?;
        Ok(value)
    }
}

#[derive(Clone, Debug, JsonSchema, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct Span {
    pub schema_version: SchemaVersion,
    pub document_kind: SpanDocumentKind,
    pub invocation_id: InvocationId,
    pub trace_id: TraceId,
    pub span_id: SpanId,
    pub parent_span_id: Option<SpanId>,
    pub name: NonBlankString,
    pub span_kind: NonBlankString,
    pub status: SpanStatus,
    pub status_message: Option<NonBlankString>,
    pub started_at: Option<Timestamp>,
    pub ended_at: Option<Timestamp>,
    pub duration_ms: Option<u64>,
    pub input_recorded: bool,
    pub input: Option<Value>,
    pub output_recorded: bool,
    pub output: Option<Value>,
    pub attributes: JsonObject,
}

impl Span {
    pub fn validate(&self) -> Result<(), &'static str> {
        if matches!(self.status, SpanStatus::Error) != self.status_message.is_some() {
            return Err("status_message contradicts span status");
        }
        if !self.input_recorded && self.input.is_some() {
            return Err("input present when input_recorded is false");
        }
        if !self.output_recorded && self.output.is_some() {
            return Err("output present when output_recorded is false");
        }
        Ok(())
    }
}

serialize_validated!(Span {
    schema_version,
    document_kind,
    invocation_id,
    trace_id,
    span_id,
    parent_span_id,
    name,
    span_kind,
    status,
    status_message,
    started_at,
    ended_at,
    duration_ms,
    input_recorded,
    input,
    output_recorded,
    output,
    attributes,
});

#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct RawSpan {
    schema_version: SchemaVersion,
    document_kind: SpanDocumentKind,
    invocation_id: InvocationId,
    trace_id: TraceId,
    span_id: SpanId,
    #[serde(deserialize_with = "required_option")]
    parent_span_id: Option<SpanId>,
    name: NonBlankString,
    span_kind: NonBlankString,
    status: SpanStatus,
    #[serde(deserialize_with = "required_option")]
    status_message: Option<NonBlankString>,
    #[serde(deserialize_with = "required_option")]
    started_at: Option<Timestamp>,
    #[serde(deserialize_with = "required_option")]
    ended_at: Option<Timestamp>,
    #[serde(deserialize_with = "required_option")]
    duration_ms: Option<u64>,
    input_recorded: bool,
    #[serde(deserialize_with = "required_option")]
    input: Option<Value>,
    output_recorded: bool,
    #[serde(deserialize_with = "required_option")]
    output: Option<Value>,
    attributes: JsonObject,
}

impl<'de> Deserialize<'de> for Span {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let raw = RawSpan::deserialize(deserializer)?;
        let value = Self {
            schema_version: raw.schema_version,
            document_kind: raw.document_kind,
            invocation_id: raw.invocation_id,
            trace_id: raw.trace_id,
            span_id: raw.span_id,
            parent_span_id: raw.parent_span_id,
            name: raw.name,
            span_kind: raw.span_kind,
            status: raw.status,
            status_message: raw.status_message,
            started_at: raw.started_at,
            ended_at: raw.ended_at,
            duration_ms: raw.duration_ms,
            input_recorded: raw.input_recorded,
            input: raw.input,
            output_recorded: raw.output_recorded,
            output: raw.output,
            attributes: raw.attributes,
        };
        value.validate().map_err(de::Error::custom)?;
        Ok(value)
    }
}

#[derive(Clone, Debug, JsonSchema, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct CheckResult {
    pub schema_version: SchemaVersion,
    pub document_kind: CheckResultDocumentKind,
    pub id: CheckResultId,
    pub invocation_id: InvocationId,
    pub name: NonBlankString,
    pub status: CheckStatus,
    pub started_at: Option<Timestamp>,
    pub ended_at: Option<Timestamp>,
    pub duration_ms: Option<u64>,
    pub evidence: JsonObject,
    pub failure: Option<Failure>,
}

impl CheckResult {
    pub fn validate(&self) -> Result<(), &'static str> {
        let failure_required = matches!(
            self.status,
            CheckStatus::Fail | CheckStatus::Error | CheckStatus::Skipped
        );
        if failure_required == self.failure.is_some() {
            Ok(())
        } else {
            Err("failure contradicts check result status")
        }
    }
}

serialize_validated!(CheckResult {
    schema_version,
    document_kind,
    id,
    invocation_id,
    name,
    status,
    started_at,
    ended_at,
    duration_ms,
    evidence,
    failure,
});

#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct RawCheckResult {
    schema_version: SchemaVersion,
    document_kind: CheckResultDocumentKind,
    id: CheckResultId,
    invocation_id: InvocationId,
    name: NonBlankString,
    status: CheckStatus,
    #[serde(deserialize_with = "required_option")]
    started_at: Option<Timestamp>,
    #[serde(deserialize_with = "required_option")]
    ended_at: Option<Timestamp>,
    #[serde(deserialize_with = "required_option")]
    duration_ms: Option<u64>,
    evidence: JsonObject,
    #[serde(deserialize_with = "required_option")]
    failure: Option<Failure>,
}

impl<'de> Deserialize<'de> for CheckResult {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let raw = RawCheckResult::deserialize(deserializer)?;
        let value = Self {
            schema_version: raw.schema_version,
            document_kind: raw.document_kind,
            id: raw.id,
            invocation_id: raw.invocation_id,
            name: raw.name,
            status: raw.status,
            started_at: raw.started_at,
            ended_at: raw.ended_at,
            duration_ms: raw.duration_ms,
            evidence: raw.evidence,
            failure: raw.failure,
        };
        value.validate().map_err(de::Error::custom)?;
        Ok(value)
    }
}

#[derive(Clone, Debug, Deserialize, JsonSchema, PartialEq, Serialize)]
#[serde(untagged)]
pub enum ProtocolDocument {
    EvalRun(Box<EvalRun>),
    Invocation(Box<Invocation>),
    Span(Box<Span>),
    CheckResult(Box<CheckResult>),
}

impl ProtocolDocument {
    pub fn document_kind(&self) -> &'static str {
        match self {
            Self::EvalRun(_) => "eval_run",
            Self::Invocation(_) => "invocation",
            Self::Span(_) => "span",
            Self::CheckResult(_) => "check_result",
        }
    }
}
